### PR TITLE
Implement screenshot test for flutter web.

### PR DESCRIPTION
**Pull request by chingjun**
## Description

 * Added a class `WebGoldenComparator` (corresponds to `GoldenFileComparator` for non-web), with default implementation being `DefaultWebGoldenComparator`.
 * `DefaultWebGoldenComparator` will issue a web request to the test server for taking and comparing the screenshot.
 * Added new files `_matchers_io.dart` and `_matcher_web.dart`, that contain the implementation of `MatchesGoldenFile`. On web, this matcher will first render the requested Element on screen, and defer the screenshot taking to the test server.
 * In `flutter_test_platform.dart`, add a `flutter_golden` handler for the test server, which does two things: screenshot taking and image comparison.
 * Screenshot is taken using WIP, and is clipped to only contain the Element requested.
 * For image comparison, it will start a `flutter_tester` process that loads `flutter_test_config.dart` file and uses the default `goldenFileComparator` to compare the golden images.
 * An additional `FLUTTER_TEST_BROWSER` environment variable is set when starting the `flutter_tester` for image comparison. `SkiaClient` is also updated to include this as one of the parameters.
 * Currently only one file is enabled for Flutter Web screenshot test, just to test the overall workflow. A followup PR will be sent to enable all screenshot tests for Flutter Web.

## Related Issues

https://github.com/flutter/flutter/issues/40297

### CODE REVIEW

File: packages/flutter_test/lib/src/_goldens_io.dart
Comment (jonahwilliams) [inline packages/flutter_test/lib/src/_goldens_io.dart]:
```suggestion
    throw UnsupportedError('DefaultWebGoldenComparator is not supported on the web.');
```

File: packages/flutter_test/lib/src/_goldens_io.dart
Comment (jonahwilliams) [inline packages/flutter_test/lib/src/_goldens_io.dart]:
```suggestion
    throw UnsupportedError('DefaultWebGoldenComparator is not supported on the web.');
```

File: packages/flutter_tools/lib/src/test/flutter_web_platform.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/test/flutter_web_platform.dart]: This code should handle the errors thrown by WIP, like WIPError (there might be some more). probably printError and move on if possible

File: packages/flutter_tools/lib/src/test/flutter_web_platform.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/test/flutter_web_platform.dart]: This is a really interesting approach. i think this PR should have at least one end to end test that goes through this path, if possible.

File: packages/flutter_goldens_client/lib/skia_client.dart
Comment (Piinks) [inline packages/flutter_goldens_client/lib/skia_client.dart]: The documentation above will need to be updated.

File: packages/flutter/test/widgets/shadow_test.dart
Comment (Piinks) [inline packages/flutter/test/widgets/shadow_test.dart]: >[ +138 ms] Preparing to send command: {"imageFile":"/tmp/flutter_web_platform.SVNUSH/imageXIEJHM/image","key":"shadow.BoxDecoration.enabled.png","update":false}
[   +1 ms] <<< Skipping "shadow.BoxDecoration.enabled.png" test : Golden file testing is unavailable on LUCI and  some Cirrus shards.

This is from the `web_tests-5-linux` shard.  I think you will need to add the gold service account, etc. to the shard that is executing the tests you would like to go through Gold. It looks like they are still skipping. `.cirrus.yml` configures the shards that use Gold: https://github.com/flutter/flutter/blob/6b11f18bc0c5a53753d1dc92928ec8ba1bdc5f95/.cirrus.yml#L90-L93

File: packages/flutter_test/lib/src/_goldens_io.dart
Comment (chingjun) [reply packages/flutter_test/lib/src/_goldens_io.dart]: Updated. It's actually `only supported` instead of `not supported though.

File: packages/flutter_test/lib/src/_goldens_io.dart
Comment (chingjun) [reply packages/flutter_test/lib/src/_goldens_io.dart]: Done.

File: packages/flutter_tools/lib/src/test/flutter_web_platform.dart
Comment (chingjun) [reply packages/flutter_tools/lib/src/test/flutter_web_platform.dart]: Done, I've handled it by catching WIPError and returning the error to the caller.

File: packages/flutter_goldens_client/lib/skia_client.dart
Comment (chingjun) [reply packages/flutter_goldens_client/lib/skia_client.dart]: Done.

File: packages/flutter/test/widgets/shadow_test.dart
Comment (chingjun) [reply packages/flutter/test/widgets/shadow_test.dart]: Updated `.cirrus.yml`. Another thing to note is that even when the test is running, it would still fail because in `flutter_goldens`, the "no baseline" case[1] will only be hit when it's a new test being added.

In this case, it's not a new test, but a new "platform" instead. Do you think it's ok for us to ignore the failure in this PR, or should I fix the part in `flutter_goldens` as well?

[1]: https://github.com/flutter/flutter/blob/master/packages/flutter_goldens/lib/flutter_goldens.dart#L513

Comment (Piinks) [reply packages/flutter/test/widgets/shadow_test.dart]: I don't think ignoring the failures will go well. 

Since this will be a huge change to many golden files, it may be worth it to wait on #44474, which is just waiting on https://chromium-review.googlesource.com/c/infra/infra/+/1933027. This will enable tryjobs. You'll be able to see all of the files you generated in pre-submit checks and bulk triage them all at once so you'll land on green. 

Otherwise, you would have to use the current workflow that will require an ignore rule to be put in place for every test until you can land and triage - which is a bit tedious as it is designed against making bulk changes.

Comment (Piinks) [reply packages/flutter/test/widgets/shadow_test.dart]: I noticed also that all of the golden file tests are now failing on that shard, does the `skip: isBrowser` not work for the other tests?

**Review by yjbanov — APPROVED** [review]
LGTM if LGT @Piinks and @jonahwilliams

File: packages/flutter/test/widgets/shadow_test.dart
Comment (chingjun) [reply packages/flutter/test/widgets/shadow_test.dart]: I've only enabled one test file with `matchesGoldenFile` right now, so only one file should be failing. Do you see other failures?

The reason it is failing is because it wasn't considered as "baseline not found". I've updated the PR with new changes in `flutter_goldens.dart` to consider the cases here as "baseline not found" as well, so the tests should pass now.

PTAL, thanks.

File: packages/flutter_tools/lib/src/test/flutter_web_platform.dart
Comment (chingjun) [reply packages/flutter_tools/lib/src/test/flutter_web_platform.dart]: Added tests for `TestGoldenComparator` and `TestGoldenComparatorProcess`, PTAL, thanks.

File: packages/flutter_goldens/lib/flutter_goldens.dart
Comment (Piinks) [inline packages/flutter_goldens/lib/flutter_goldens.dart]: This may not go over well. 

The reason why expectation parameters are _only_ validated after a match is not found is because the Gold `/details` api is expensive and subject to a rate limit. The request may timeout if you check the `/details` like this for _every_ test and _every_ expectation. It would also mean a lot of unnecessary requests.

If there is a matching image from the given expectations, then checking the `/details` is irrelevant, so they are checked only when a match cannot be found in the current implementation.

cc/ @kjlubick

File: packages/flutter_test/lib/src/matchers.dart
Comment (Piinks) [inline packages/flutter_test/lib/src/matchers.dart]: Is there potential for confusion between `matchesGoldenFile` and `MatchesGoldenFile` now that this will be publicly exposed?

File: packages/flutter_test/lib/src/_matchers_io.dart
Comment (Piinks) [inline packages/flutter_test/lib/src/_matchers_io.dart]: Are there ramifications to exposing this publicly as well?

**Review by Piinks — COMMENTED** [review]
Also, this appears to be breaking the current pre-submit workflow, see test failure in the framework-misc shards.

File: packages/flutter_goldens/lib/flutter_goldens.dart
Comment (kjlubick) [reply packages/flutter_goldens/lib/flutter_goldens.dart]: Yeah, check (and cache) the baseline from /json/expectations, since that will have all known digests for all tests and is very cheap (much higher rate limit).

**Comment by chingjun** [comment]
Talked to @Piinks and we are going to wait on #44474, reverted part of my changes in `flutter_golden`.

File: packages/flutter_goldens/lib/flutter_goldens.dart
Comment (chingjun) [reply packages/flutter_goldens/lib/flutter_goldens.dart]: As mentioned in another comment, we are going to wait for TryJobs for this, reverted the changes here.

File: packages/flutter_test/lib/src/matchers.dart
Comment (chingjun) [reply packages/flutter_test/lib/src/matchers.dart]: This is still not "publicly exposed" bacause it's not exported from `flutter_test.dart`

We moved platform-specific implementations into a separate file and use conditional imports to conditionally import the files, which is why we had to make it "public".

cc @jonahwilliams , are there other best practices for this scenario?

Comment (jonahwilliams) [reply packages/flutter_test/lib/src/matchers.dart]: Generally the way I've been doing it is:

```
_library_io.dart
_library_web.dart
library.dart (with conditional import/export of the above)
```

This makes it extra obvious that one should not import the _io or _web libraries too

**Comment by Piinks** [comment]
#44474 has landed. If you update your branch, you should see the results of pre-submit checks on the [Flutter Gold dashboard](https://flutter-gold.skia.org/?query=source_type%3Dflutter) under Changelists, where you can approve them ahead of time.

File: packages/flutter_test/lib/src/matchers.dart
Comment (Piinks) [reply packages/flutter_test/lib/src/matchers.dart]: sgtm!

**Comment by chingjun** [comment]
Thanks Kate! Just rebased and updated my branch, waiting for the tests to run :)

**Comment by Piinks** [comment]
Not all of the tests have finished yet, but it is looking promising!

Screen Shot 2019-12-04 at 4 33 22 PM

File: packages/flutter/test/widgets/shadow_test.dart
Comment (Piinks) [inline packages/flutter/test/widgets/shadow_test.dart]: Why is this removed?

File: packages/flutter/test/widgets/shadow_test.dart
Comment (Piinks) [inline packages/flutter/test/widgets/shadow_test.dart]: It looks like it ended up here?

File: packages/flutter/test/widgets/shadow_test.dart
Comment (Piinks) [inline packages/flutter/test/widgets/shadow_test.dart]: Same here?

**Review by Piinks — COMMENTED** [review]
Curious, does this enable golden file testing for web for everyone?

File: packages/flutter/test/widgets/shadow_test.dart
Comment (chingjun) [reply packages/flutter/test/widgets/shadow_test.dart]: You caught me in a temporary commit where I wanted to speed up triaging golden file changes.

The test aborted after the first golden failure. As you see here, there are 10 golden comparisons in a single test, if I ran the tests normally, only one of the 10 images shows up in the triage page. So instead of putting all the comparisons in a single test, I split them up temporarily into multiple tests, so that all the new images are uploaded in a single run, and I can triage all of them at once, instead of rerunning the test 10 times :)

**Comment by chingjun** [comment]
> Curious, does this enable golden file testing for web for everyone?

Yes. This PR only enabled that for a single test, and we plan to enable all the other tests in a later PR.

File: packages/flutter/test/widgets/shadow_test.dart
Comment (chingjun) [reply packages/flutter/test/widgets/shadow_test.dart]: Resolve to this since it is commenting on a temporary change.

File: packages/flutter/test/widgets/shadow_test.dart
Comment (chingjun) [reply packages/flutter/test/widgets/shadow_test.dart]: Explained in another comment. Resolving this since it is commenting on a temporary change.

File: packages/flutter/test/widgets/shadow_test.dart
Comment (chingjun) [reply packages/flutter/test/widgets/shadow_test.dart]: Explained in another comment. Resolving this since it is commenting on a temporary change.

**Comment by chingjun** [comment]
The tests now passed with the new images all approved on skia-gold.

@Piinks @jonahwilliams PTAL, thanks.

File: packages/flutter_tools/lib/src/test/flutter_web_platform.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/test/flutter_web_platform.dart]: Does this actually clip the screenshot in anyway? What if it is offset from the origin?

File: packages/flutter_tools/lib/src/test/flutter_web_platform.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/test/flutter_web_platform.dart]: catch FormatException from decoding

File: packages/flutter_tools/lib/src/test/flutter_web_platform.dart
Comment (chingjun) [reply packages/flutter_tools/lib/src/test/flutter_web_platform.dart]: We are calling `window.render()` before taking the screenshot (in `_matchers_web.dart`), so the element is supposed to be at the origin.

https://github.com/flutter/flutter/pull/45530/files#diff-a337fbcec47632d587865bec77ec07ddR88

File: packages/flutter_tools/lib/src/test/flutter_web_platform.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/test/flutter_web_platform.dart]: Does this work in the presence of test concurrency? Or does that not work for golden files?

File: packages/flutter_tools/lib/src/test/flutter_web_platform.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/test/flutter_web_platform.dart]:
```suggestion
        .map<dynamic>(jsonDecode)
```

File: packages/flutter_tools/lib/src/test/runner.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/test/runner.dart]: Maybe something like "Cannot execute tester at $shellPath"

File: packages/flutter_tools/test/general.shard/web/golden_comparator_test.dart
Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/web/golden_comparator_test.dart]:
```suggestion
    test('succeed when golden comparison succeed', () => testbed.run(() async {
```

File: packages/flutter_tools/test/general.shard/web/golden_comparator_test.dart
Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/web/golden_comparator_test.dart]:
```suggestion
    test('fail with error message when golden comparison failed', () => testbed.run(() async {
```

File: packages/flutter_tools/test/general.shard/web/golden_comparator_test.dart
Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/web/golden_comparator_test.dart]:
```suggestion
    test('does not reuse the process for different test file', () => testbed.run(() async {
```

**Review by jonahwilliams — APPROVED** [review]
Just some small nits, overall LGTM

File: packages/flutter_tools/lib/src/test/flutter_web_platform.dart
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/test/flutter_web_platform.dart]: Can you document that here and where the render is done for posterity?

Comment (chingjun) [reply packages/flutter_tools/lib/src/test/flutter_web_platform.dart]: Added comment here, and in `_matches_web.dart` as well.

File: packages/flutter_tools/lib/src/test/flutter_web_platform.dart
Comment (chingjun) [reply packages/flutter_tools/lib/src/test/flutter_web_platform.dart]: Done

File: packages/flutter_tools/lib/src/test/flutter_web_platform.dart
Comment (chingjun) [reply packages/flutter_tools/lib/src/test/flutter_web_platform.dart]: Done.

File: packages/flutter_tools/lib/src/test/runner.dart
Comment (chingjun) [reply packages/flutter_tools/lib/src/test/runner.dart]: Done. I actually copied this text from below, in the same file :)

File: packages/flutter_tools/test/general.shard/web/golden_comparator_test.dart
Comment (chingjun) [reply packages/flutter_tools/test/general.shard/web/golden_comparator_test.dart]: Done.

File: packages/flutter_tools/test/general.shard/web/golden_comparator_test.dart
Comment (chingjun) [reply packages/flutter_tools/test/general.shard/web/golden_comparator_test.dart]: Done.

File: packages/flutter_tools/test/general.shard/web/golden_comparator_test.dart
Comment (chingjun) [reply packages/flutter_tools/test/general.shard/web/golden_comparator_test.dart]: Done.

File: packages/flutter_tools/lib/src/test/flutter_web_platform.dart
Comment (chingjun) [reply packages/flutter_tools/lib/src/test/flutter_web_platform.dart]: I noticed that we have `--concurrency=1` for web tests on cirrus so I didn't take that into account when writing this.

Do we actually run tests concurrently for web? If so I can try to fix this.

Comment (chingjun) [reply packages/flutter_tools/lib/src/test/flutter_web_platform.dart]: Just talked to @yjbanov , the test runner actually uses the same browser to run multiple tests in multiple iframes, so screenshot taking doesn't work in this case.

**Review by Piinks — APPROVED** [review]
LGTM! 🎉 📸

**Comment by chingjun** [comment]
Thank you everyone who has reviewed and given feedback to this PR!

**Pull request merged**
